src/Domain/Block.cpp

```diff
@@ -61,48 +62,23 @@ bool operator!=(const Block<VolumeDim, TargetFrame>& lhs,
   return not(lhs == rhs);
 }
 
-template class Block<1, Frame::Grid>;
-template class Block<2, Frame::Grid>;
-template class Block<3, Frame::Grid>;
-template class Block<1, Frame::Inertial>;
-template class Block<2, Frame::Inertial>;
-template class Block<3, Frame::Inertial>;
+#define GET_DIM(data) BOOST_PP_TUPLE_ELEM(0, data)
+#define GET_FRAME(data) BOOST_PP_TUPLE_ELEM(1, data)
 
-template std::ostream& operator<<(std::ostream& os,
-                                  const Block<1, Frame::Grid>& block);
-template std::ostream& operator<<(std::ostream& os,
-                                  const Block<2, Frame::Grid>& block);
-template std::ostream& operator<<(std::ostream& os,
-                                  const Block<3, Frame::Grid>& block);
-template std::ostream& operator<<(std::ostream& os,
-                                  const Block<1, Frame::Inertial>& block);
-template std::ostream& operator<<(std::ostream& os,
-                                  const Block<2, Frame::Inertial>& block);
-template std::ostream& operator<<(std::ostream& os,
-                                  const Block<3, Frame::Inertial>& block);
+#define INSTANTIATION(r, data)                                               \
```

Comment:
What is `r`?

Reply:
This is part of Boost.Preprocessor: https://www.boost.org/doc/libs/1_68_0/libs/preprocessor/doc/index.html

Reply:
What is that difference to using an underscore instead of "r" like everywhere else in the code? If there is no difference, the underscore should be used for consistency.

Reply:
Different characters in the alphabet, that's it. `r` is consistent with Boost. `_` is usually used in the three argument macros